lib/Common/Common/CompressionUtilities.cpp

```diff
@@ -0,0 +1,138 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "CommonCommonPch.h"
+#include "CompressionUtilities.h"
+
+#ifdef ENABLE_COMPRESSION_UTILITIES
+#include <compressapi.h>
+#endif
+
+#define IFFALSEGO(expr,label) do { if(!(expr)) { goto label; } } while(0);
+
+using namespace Js;
+
+DWORD ConvertCompressionAlgorithm(CompressionUtilities::CompressionAlgorithm algorithm)
+{
+    // Note: The algorithms listed in CompressionUtilities.h should be kept in-sync with those
+    // defined in compressapi.h or else we will need to do more than a simple cast here.
+    return static_cast<DWORD>(algorithm);
+}
+
+HRESULT CompressionUtilities::CompressBuffer(
+    _In_ ArenaAllocator* alloc,
+    _In_ const byte* inputBuffer,
+    _In_ size_t inputBufferByteCount,
+    _Out_ byte** compressedBuffer,
+    _Out_ size_t* compressedBufferByteCount,
+    _In_opt_ CompressionAlgorithm algorithm)
+{
+    Assert(compressedBuffer != nullptr);
+    Assert(compressedBufferByteCount != nullptr);
+
+    *compressedBuffer = nullptr;
+    *compressedBufferByteCount = 0;
+
+    HRESULT hr = E_FAIL;
+
+#ifdef ENABLE_COMPRESSION_UTILITIES
+    COMPRESSOR_HANDLE compressor = nullptr;
+    if (!CreateCompressor(ConvertCompressionAlgorithm(algorithm), nullptr, &compressor))
+    {
+        return E_FAIL;
```

Comment:
Do we necessarily want to return E_FAIL here instead of getting an HRESULT from GetLastError? I'm not sure what the standard is (to return detailed errors from just the function that's effectively being wrapped here, or to do it for everything).

Reply:
There isn't really a standard way we go about this. Below I did grab the error code from GetLastError and translate it to an HRESULT because there's an error code which is returned via Compress/Decompress which is actually a success condition when you just want to find out the size of the output buffer without doing the operation.

I'll try and go through getting the last error for failure cases.